### PR TITLE
Adding HeapedVector.jl 

### PR DESCRIPTION
**Pull request by yashcodes**
solves [Issue#6](https://github.com/JuliaIntervals/IntervalOptimisation.jl/issues/6)

### CODE REVIEW

File: benchmark/benchmarks.jl [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using .OneDimExamples, .TwoDimExamples
+
+using BenchmarkTools
+using IntervalArithmetic
+using IntervalOptimisation
+
+const SUITE = BenchmarkGroup()
+
+
+S = SUITE["One dimentional function"] = BenchmarkGroup()
```
Comment (dpsanders) [inline benchmark/benchmarks.jl]: dimensional (t->s)

File: benchmark/benchmarks.jl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+for func in (OneDimExamples.func1, OneDimExamples.func2, OneDimExamples.func3)
+    s = S[func.title] = BenchmarkGroup()
+	for Structure in (HeapedVector, SortedVector)
```
Comment (dpsanders) [inline benchmark/benchmarks.jl]: structure (S->s)

File: benchmark/benchmarks.jl [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+		s[string(Structure)] = @benchmarkable minimise($(func.func), $(func.region), structure = $Structure)
+    end
+end    
+
+
+S = SUITE["Two dimentional function"] = BenchmarkGroup()
```
Comment (dpsanders) [inline benchmark/benchmarks.jl]: s

File: benchmark/examples.jl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+using IntervalArithmetic, IntervalOptimisation
+
+struct working_function{string, F<:Function, reg<:Interval} 
+    func::F
+    region::reg
+    title::string
+end
+
+func1 = working_function("Quadrtic equation", Interval(-1,2), x->x*(x-1) )
```
Comment (dpsanders) [inline benchmark/examples.jl]: quadratic

File: benchmark/examples.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+func2 = working_function("Cubic equation", Interval(-1,4), x->(x-2)^3-5x )
+func3 = working_function("quadritic equation", Interval(-4,6), x->(x-6)*(x+4)*(7x^2+10x+24) )
```
Comment (dpsanders) [inline benchmark/examples.jl]: quadratic

File: benchmark/examples.jl (hunk 4 above)
Comment (dpsanders) [inline benchmark/examples.jl]: -1..2

File: benchmark/examples.jl [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+end 
+
+
+
+module TwoDimExamples
+
+using IntervalArithmetic, IntervalOptimisation
+
+struct working_function{string, F<:Function, reg<:Union{Interval, IntervalBox}}
```
Comment (dpsanders) [inline benchmark/examples.jl]: `WorkingFunction`

`struct`s have initial capitals.

Do not parametrise by `string`

File: benchmark/examples.jl [hunk 7]
```diff
@@ -0,0 +1,46 @@
+module OneDimExamples
+
+using IntervalArithmetic, IntervalOptimisation
+
+struct working_function{string, F<:Function, reg<:Interval}
```
Comment (dpsanders) [inline benchmark/examples.jl]: It doesn't  seem necessary to make a type for this?

File: benchmark/examples.jl [hunk 8]
```diff
@@ -0,0 +1,46 @@
+module OneDimExamples
+
+using IntervalArithmetic, IntervalOptimisation
+
+struct working_function{string, F<:Function, reg<:Interval} 
+    func::F
+    region::reg
```
Comment (dpsanders) [inline benchmark/examples.jl]: region::Interval

File: benchmark/examples.jl [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    title::string
```
Comment (dpsanders) [inline benchmark/examples.jl]: title::String

File: benchmark/examples.jl (hunk 6 above)
Comment (dpsanders) [inline benchmark/examples.jl]: This can be defined only once, outside the modules.

File: src/HeapedVectors.jl [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+import Base: getindex, length, push!, isempty,
+        pop!, filter!, popfirst!
+
+export HeapedVector
+
+struct HeapedVector{T, F<:Function}  
+    data::Vector{T}
+    by::F
+    function HeapedVector(v::Vector{T}, by::F) where {T, F}
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: Add a blank line before this function definition.

File: src/HeapedVectors.jl [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    end
+
+    if smallest != index
+       v.data[index], v.data[smallest] = v.data[smallest], v.data[index]
+	   bubbledown!(v, smallest)
+    end
+end
+
+
+function filter!(A::HeapedVector{T}, x::T) where{T}
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: This is not the correct use of the `filter!` function. It should have a different name.

File: test/heaped_vector.jl [hunk 12]
```diff
@@ -0,0 +1,36 @@
+using IntervalOptimisation
+using Test
+
+const HeapedVector = IntervalOptimisation.HeapedVector
+
+@testset "heap" begin
+	
+	@testset "with standard heaping " begin
```
Comment (dpsanders) [inline test/heaped_vector.jl]: Indentation

File: test/optimise.jl [hunk 13]
```diff
@@ -1,42 +1,62 @@
...
         @test minimisers[1] ⊆ -0.1..0.1
 
         global_min, minimisers = minimise(x->(x^2-2)^2, -10..11)
         @test global_min ⊆ 0..1e-7
         @test length(minimisers) == 2
-        @test sqrt(2) ∈ minimisers[1]
+        @test sqrt(2) ∈ minimisers[2]
     end
 
+    for Structure in (SortedVector, HeapedVector)
```
Comment (dpsanders) [inline test/optimise.jl]: structure

File: test/optimise.jl [hunk 14]
```diff
@@ -1,42 +1,62 @@
...
 
-    @testset "Discontinuous function in 1D" begin
+        @testset "Minimise in 1D using SoretedVector" begin
```
Comment (dpsanders) [inline test/optimise.jl]: This is still not correct. 
```
"using $structure"
```

File: src/HeapedVectors.jl [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+	   smallest = right
+    end
+
+    if smallest != index
+       v.data[index], v.data[smallest] = v.data[smallest], v.data[index]
+	   bubbledown!(v, smallest)
+    end
+end
+
+function filter_elements!(A::HeapedVector{T}, x::T) where{T}
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: You need to pass `func` as an argument to this function.

File: src/HeapedVectors.jl [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    func(y) = A.by(y) < A.by(x)
+    filter!(func, A.data)
+
+    if length(A.data) == 0
+        return A
+    end
+
+    ar = heaping(A.data, A.by)
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: So you just reconstruct the whole heap each time? That seems inefficient.

Comment (yashcodes) [reply src/HeapedVectors.jl]: Actually after deleting all the elements which should be deleted. The remaining elements are arranged in heap again recursively. It is necessary to keep it work in Heap in subsequent steps.

Comment (yashcodes) [reply src/HeapedVectors.jl]: In any approach we need to keep all the elements in heaped form so we need to do this.

Comment (dpsanders) [reply src/HeapedVectors.jl]: But you could do it by deleting elements one by one instead, which may be more efficient (or not). You can also just not bother filtering the elements at all!

Comment (yashcodes) [reply src/HeapedVectors.jl]: one thing we can do that replace 
`ar = heaping(A.data, A.by)`
by
`A.data = heaping(A.data, A.by)`

and delete for loop below it

Comment (dpsanders) [reply src/HeapedVectors.jl]: I think you need `.=`

File: src/HeapedVectors.jl [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+function filter_elements!(A::HeapedVector{T}, x::T) where{T}
+    func(y) = A.by(y) < A.by(x)
+    filter!(func, A.data)
+
+    if length(A.data) == 0
+        return A
+    end
+
+    ar = (heaping(A.data, A.by))
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: Too many parentheses

File: src/HeapedVectors.jl [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+	convert(typeof(ar), A.data)
```
Comment (dpsanders) [inline src/HeapedVectors.jl]: This line has no effect. What are you trying to do?

File: src/SortedVectors.jl [hunk 19]
```diff
@@ -3,31 +3,29 @@ __precompile__()
 module SortedVectors
 
 import Base: getindex, length, push!, isempty,
-        pop!, resize!, popfirst!
+        pop!, filter!, popfirst!
```
Comment (dpsanders) [inline src/SortedVectors.jl]: No need to import `filter!`

File: src/HeapedVectors.jl (hunk 18 above)
Comment (yashcodes) [reply src/HeapedVectors.jl]: Yes you are right. I was thinking something else

**Comment by dpsanders** [comment]
Thanks a lot for the contribution!

**Pull request merged**